sigstore/model-transparency
# Prepare for wheel publishing (alpha patch version of 0.0.1)

#### Summary
Prepare the project so that we can publish a wheel with the library (as part of #175). Attempt to make sure we can still proceed with both building the library and the surrounding CLI.

We are going to use `hatch` for all development work as that makes it very simple:
- to test: `hatch test`
- to test with coverage: `hatch test -c`
- to test against all supported python versions: `hatch test --all` (`-c` still supported)
- to format the code: `hatch fmt`
- to check the formatting (in GHA, etc.): `hatch fmt --check`.
- to check typing: `hatch run type:check`. This is a custom environment and we can create others in the future (e.g., for documentation)
- to build the wheel: `hatch build`. The wheel can then be installed in a separate virtual env if needed for future testing.
- to run custom tests: `hatch shell` or `hatch run $cmd`.

Testing now happens in parallel, and in randomized fashion. Fast and removes the possibility of a test case influencing the behavior of another.

Another benefit of using `hatch` is that it allows us to build C/C++ extensions (via plugins), so when we need to get extra performance or drop to C/C++ (e.g., for #201) we can just do it.

---

In order to make this buildable as a wheel, I needed to restructure the code. So, the first commit moves files around and only changes 2 markdown links. The files are moved to get to a [`src/` layout](https://packaging.python.org/en/latest/discussions/src-layout-vs-flat-layout/#src-layout-vs-flat-layout). There are more benefits for a `src/` layout than a flat
layout, such as:

- documentation is easier to generate
- we won't accidentally test against code that is different than what users would see
- we can include only the source in the wheel, reducing PyPI bandwidth
- it makes it clearer which parts of the repo go into the wheel and which ones are ancillary. Incidentally, we might now want to move the SLSA parts to another repo and maybe rename this one? We can do this at a later time.

Since we would need to change `README.model_signing.md` anyway once the API is ready, I have not updated it besides just fixing the image pointer.

I have deleted old files from when this was a POC. Since we're working on releasing the library, we no longer need those. The code in these files is already implemented in the library. What these files used to test is now in the test files.

The alternative to not moving the files to the root of the folder would have been to duplicate `LICENSE`, etc. inside the `model-signing` directory, so that packaging tools would see it. But we still needed to move files around to get to the `src/` layout and to make documentation work correctly. So, it did not make sense to also duplicate `LICENSE`.

---

For the second commit of the PR I just focused on making sure that `hatch build` builds the wheel and the tarball. As a bonus, `hatch shell` drops us into a virtual environment with the package locally installed, in an editable fashion: we can just edit the files and can immediately test the code without having to reinstall the wheel somewhere else. So, we get the benefits of testing exactly the wheel that is shipped to the users without having to have a separate virtual environment to install the wheel in, without having to manually manage that.

The wheel that gets built is minimal. Here's the corresponding tarball's contents (output of `tar tf`):

```
model_signing-0.0.1a0/model_signing/__init__.py
model_signing-0.0.1a0/model_signing/hashing/__init__.py
model_signing-0.0.1a0/model_signing/hashing/file.py
model_signing-0.0.1a0/model_signing/hashing/hashing.py
model_signing-0.0.1a0/model_signing/hashing/memory.py
model_signing-0.0.1a0/model_signing/manifest/__init__.py
model_signing-0.0.1a0/model_signing/manifest/manifest.py
model_signing-0.0.1a0/model_signing/serialization/__init__.py
model_signing-0.0.1a0/model_signing/serialization/serialization.py
model_signing-0.0.1a0/model_signing/serialization/serialize_by_file.py
model_signing-0.0.1a0/model_signing/serialization/serialize_by_file_shard.py
model_signing-0.0.1a0/model_signing/signature/__init__.py
model_signing-0.0.1a0/model_signing/signature/encoding.py
model_signing-0.0.1a0/model_signing/signature/fake.py
model_signing-0.0.1a0/model_signing/signature/key.py
model_signing-0.0.1a0/model_signing/signature/pki.py
model_signing-0.0.1a0/model_signing/signature/signing.py
model_signing-0.0.1a0/model_signing/signature/sigstore.py
model_signing-0.0.1a0/model_signing/signature/verifying.py
model_signing-0.0.1a0/model_signing/signing/__init__.py
model_signing-0.0.1a0/model_signing/signing/as_bytes.py
model_signing-0.0.1a0/model_signing/signing/empty_signing.py
model_signing-0.0.1a0/model_signing/signing/in_toto.py
model_signing-0.0.1a0/model_signing/signing/signing.py
model_signing-0.0.1a0/model_signing/signing/sigstore.py
model_signing-0.0.1a0/.gitignore
model_signing-0.0.1a0/LICENSE
model_signing-0.0.1a0/README.model_signing.md
model_signing-0.0.1a0/pyproject.toml
model_signing-0.0.1a0/PKG-INFO
```

---

I added a few more formatting settings and run `hatch fmt` across the entire code, affecting `slsa_for_models` too. Added some excludes and we'll work on minimizing those in the future.

---

Updated `.gitignore` to exclude all the artifacts generated during development. Updated GitHub Actions to run via `hatch`. Removed GHA workflows which became obsolete, as well as the requirements files for `model_signing`, since now dependencies are handled via `hatch`.

---

Since CodeQL was complaining about `signature/sigstore.py` and we already have `singing/sigstore.py` with the implementation of signing with Sigstore, I removed the file in `signature`. Incidentally, we might want to rename `signature/` as `pki/` as it only has the parts needed for signing with own PKI. We would then offer this as a feature of the wheel (`pip install model_signing[pki]`) instead of having it be there by default. But these changes will happen in future PRs.

---

This would require rebase on #279, #240 given that we have moved files around. It closes #192 given that all tests have been migrated. It closes #68 due to the same. Similarly, closes #182. Also, closes #231 since we have fixed `slsa_for_models` code to pass `hatch fmt`

#### Release Note
NONE (will add when ready to publish, also with CHANGELOG.md)

#### Documentation
NONE (will add when ready to publish, there's some TODO now to merge `README.md` files)
